zerver/lib/actions.py

```diff
@@ -4991,8 +4991,7 @@ def notify_user_update_custom_profile_data(user_profile: UserProfile,
         data["value"] = ujson.dumps(field['value'])
     else:
         data['value'] = field['value']
-    if field['rendered_value']:
-        data['rendered_value'] = field['rendered_value']
+    data['rendered_value'] = field['rendered_value']
```

Comment:
It looks like this moves us back to always sending a `rendered_value`, even if it's None and/or the field type doesn't support rendered values; what's your thinking here?  

(Same behavior is present in the `events.py` code; we should either have if statements in these places or not?)

Reply:
I guess I should have mentioned reverting back to sending rendered_value for all fields. Sorry about not mentioning it before. Here's my thinking explained via. an example case:

Take this example from the client end:
```javascript
exports.set_custom_profile_field_data = function (user_id, field) {
    if (field.id === undefined) {
        blueslip.error("Unknown field id " + field.id);
        return;
    }
    people_by_user_id_dict.get(user_id).profile_data[field.id] = {
        value: field.value,
        rendered_value: field.rendered_value,
    };
};
```

If we don't want to send the rendered_value during communication between the server and client ends when not absolutely required then that can mean: 
- we want to optimize the size of data transmitted by not sending the one extra key
- we want to do the above but then also save some memory on the client end by not storing the rendered value.

Either case will lead to bulkier code with what I ***think*** would be some nominal amount of optimization.

in the first case the example given above would become:
```javascript
exports.set_custom_profile_field_data = function (user_id, field) {
    if (field.id === undefined) {
        blueslip.error("Unknown field id " + field.id);
        return;
    }
    if (field.type === field_types.LONG_TEXT.id || field.type === field_types.SHORT_TEXT.id) {
        people_by_user_id_dict.get(user_id).profile_data[field.id] = {
            value: field.value,
            rendered_value: field.rendered_value,
        };
    }
    else
      people_by_user_id_dict.get(user_id).profile_data[field.id] = field.value;
};
```

in the second case, the code would be changed to something like:
```javascript
exports.set_custom_profile_field_data = function (user_id, field) {
    if (field.id === undefined) {
        blueslip.error("Unknown field id " + field.id);
        return;
    }
    if (field.type === field_types.LONG_TEXT.id || field.type === field_types.SHORT_TEXT.id) {
        people_by_user_id_dict.get(user_id).profile_data[field.id] = {
            value: field.value,
            rendered_value: field.rendered_value,
        };
    }
    else
      people_by_user_id_dict.get(user_id).profile_data[field.id] = {
            value: field.value,
            rendered_value: null,
        };
};
```

While with this example, it doesn't seem like a big deal, we would have to make similar changes in other places as well, and I thought that these extra conditional statements would bulk up the code and also nominally slow it down in exchange for a nominal amount of memory saved.

e.g. In the second case (to save on memory) we would also have to change get_custom_profile_data to:
```javascript
exports.get_custom_profile_data = function (user_id, field_id) {
    var field = people_by_user_id_dict.get(user_id).profile_data[field_id];
    if (field.type === field_types.SHORT_TEXT || field.type === field_types.LONG_TEXT) {
      return field.value;
    }
    return field;
};
```
whereas it used to be
```javascript
exports.get_custom_profile_data = function (user_id, field_id) {
    return people_by_user_id_dict.get(user_id).profile_data[field_id];
```

*Note:*  
The first case means that we save nominally on data transmitted (just one key) at the expense of breaking some uniformity and introducing extra conditionals statements.  
The second case means completely breaking uniformity starting at `exports.get_custom_profile_data ()` and introducing a some extra conditional checks which would make the code a little messier to read.

Reply:
Now I don't mind reverting back to only sending the rendered_value when needed and breaking the uniformity of what `exports.get_custom_profile_data ()` returns... it's only a small change anyways. But I don't see enough of an issue with sending rendered_values for all custom profile fields. If there is a flaw in my approach here please do let me know, I'd be happy to make the change if you see it fit.

Reply:
They're both reasonable designs, but I think it clutters the API more to have a `rendered_content: None` sent for every field type that doesn't support the feature.  And in any case, you save a bit of bandwidth not sending a useless None in all fields of the types for which `rendered_content` is invalid.

Reply:
But @timabbott would it be alright if we kept the value inside a dictionary? As in, before we used to just send the value regularly but with the introduction of rendered values, I started sending the value(s) as `{"value": value, "rendered_value": rendered_value}`. So for fields that are not renderable, should we send the value as just `value` or would `{"value": value} `be fine?